rumqttc/src/lib.rs

```diff
@@ -624,11 +613,44 @@ impl MqttOptions {
     /// When set `false`, broker will hold the client state and performs pending
     /// operations on the client when reconnection with same `client_id`
     /// happens. Local queue state is also held to retransmit packets after reconnection.
+    ///
+    /// NOTE: It panicks if the `client_id` is a zero byte string and `clean_session` is
+    /// not `true`.
+    ///
+    /// ```should_panic
+    /// # use rumqttc::MqttOptions;
+    /// let mut options = MqttOptions::new("", "localhost", 1883);
+    /// options.set_clean_session(false);
+    /// ```
     pub fn set_clean_session(&mut self, clean_session: bool) -> &mut Self {
+        if self.client_id.is_empty() {
+            assert!(
+                clean_session,
+                "Cannot unset clean session when client id is empty"
+            );
+        }
         self.clean_session = clean_session;
         self
     }
 
+    /// Does the same thing as [`Self::set_clean_session`], however it returns an error
+    ///  if `clean_session = false` and `client_id` is a zero byte string.
+    ///
+    /// ```
+    /// # use rumqttc::{MqttOptions, OptionError};
+    /// let mut options = MqttOptions::new("", "localhost", 1883);
+    /// let error = options.try_set_clean_session(false).unwrap_err();
+    /// assert_eq!(error, OptionError::CleanSession);
+    /// ```
+    #[cfg(feature = "url")]
+    pub fn try_set_clean_session(&mut self, clean_session: bool) -> Result<&mut Self, OptionError> {
+        if self.client_id.is_empty() && !clean_session {
+            return Err(OptionError::CleanSession);
+        }
+        self.clean_session = clean_session;
+        Ok(self)
+    }
```

Comment:
`url` feature is only used for `from_url` fn, this fn isn't required here! 

we should add this check in url's try_from impl block!

note: there is PR to refactor MqttOptions which would remove this url feature #789